relayer/src/connection.rs

```diff
@@ -500,114 +410,231 @@ impl<ChainA: ChainHandle, ChainB: ChainHandle> Connection<ChainA, ChainB> {
+                    e
+                })?;
+            }
+
+            // send the Ack message to chain b (destination)
+            (State::TryOpen, State::Init) => {
+                self.build_conn_ack_and_send().map_err(|e| {
+                    error!("failed ConnOpenAck {:?}: {:?}", self.b_side, e);
+                    e
+                })?;
+            }
+
+            // send the Confirm message to chain b (destination)
+            (State::Open, State::TryOpen) => {
+                self.build_conn_confirm_and_send().map_err(|e| {
+                    error!("failed ConnOpenConfirm {:?}: {:?}", self.b_side, e);
+                    e
+                })?;
+            }
+
+            // send the Confirm message to chain a (source)
+            (State::TryOpen, State::Open) => {
+                self.flipped().build_conn_confirm_and_send().map_err(|e| {
+                    error!("failed ConnOpenConfirm {:?}: {:?}", self.a_side, e);
+                    e
+                })?;
+            }
+
+            (State::Open, State::Open) => {
+                info!("connection handshake already finished for {:#?}\n", self);
+                return Ok(());
+            }
+
+            (a_state, b_state) => {
+                warn!(
+                    "do_conn_open_finalize does not handle connection end state combination: \
+                    {}-{}, {}-{}. will retry to account for RPC node data availability issues.",
+                    self.a_chain().id(),
+                    a_state,
+                    self.b_chain().id(),
+                    b_state
+                );
             }
         }
+        Err(ConnectionError::handshake_finalize())
+    }
 
-        Err(ConnectionError::max_retry())
+    /// Executes the connection handshake protocol (ICS003)
+    fn handshake(&mut self) -> Result<(), ConnectionError> {
+        let max_block_times = self.max_block_times()?;
+        // retry every 1/10th of block time
+        let retry_delay = max_block_times / 10;
+        let strategy = clamp_total(
+            ConstantGrowth::new(retry_delay, Duration::from_secs(0)),
+            retry_delay,
+            max_block_times * 100, // retry for ~100 blocks
+        );
+        retry_with_index(strategy, |_| {
+            if let Err(e) = self.do_conn_open_handshake() {
+                if e.is_expired_or_frozen_error() {
+                    RetryResult::Err(e)
+                } else {
+                    RetryResult::Retry(e)
+                }
+            } else {
+                RetryResult::Ok(())
+            }
+        })
+        .map_err(|err| {
+            error!(
+                "failed to open connection after {} retries",
+                retry_count(&err)
+            );
+            from_retry_error(
+                err,
+                format!("failed to finish connection handshake for {:?}", self),
+            )
+        })?;
```

Comment:
Does this `map_err` call get triggered upon every retry, or only once the maximum number of retries has been reached?